### PR TITLE
[NPU] Support dsv4 dspark on ascend 950

### PR DESCRIPTION
**Pull request by randgun**
## Motivation

Enable DeepSeek V4 DSpark speculative decoding on NPU Ascend 950 (arch35). The existing DSpark implementation only supported CUDA and older NPU (arch22). Arch35 hardware requires different sparse-attention operators, weight-loading strategies, and numerical compatibility workarounds. Several follow-up fixes address a regression where DSpark's unquantized `wo_a` path conflicted with the arch35 MXFP8 detection logic, plus NPU-specific numerical gaps.

## Modifications

- **arch35 DSpark sparse attention**: Added arch35-specific v2 sparse attention operators via a new `TorchOpLoader` extension; loaded during TP worker and DSpark draft worker initialization.
- **`wo_a` quantization fix**: `use_npu_arch35_mxfp8_wo_a()` now accepts a `wo_a_keeps_quant_config` parameter and returns `False` when DSpark explicitly disables `wo_a` quantization. This fixes a regression introduced by a previous commit that changed the assertion condition from `if fp8:` to `if quantize_wo_a:`, which caused an incorrect assertion on arch35 because DSpark's `wo_a` was not quantized. The model also caches this value instead of recomputing it at each call site.
- **DSpark stage configuration**: Added `_dspark_stage_config()` to apply draft-specific expert counts to draft stages and disable vision layers for text-only stages. Arch35 now falls back to standard vocabulary/fusion behavior instead of the arch31-specific path.
- **Weight loading**: Moved the NPU `weight_packed` → `weight` rename before stacked parameter mapping, eliminating duplicated branch logic.
- **NPU numerical compatibility**: Fall back to torch for commit-KV projection on NPU; replaced `exponential_()` (unsupported on NPU) with `uniform_ + log_ + neg_`.

## Accuracy Tests

| Model | Dataset | Metric | Subset | Num | Score |
|---|---|---|---:|---:|---:|
| DeepSeek-V4-Flash-0731 | GPQA-Diamond | Accuracy ↑ | gpqa_diamond | 198 | 89.9% |

## Speed Tests and Profiling
| Model | Dataset | Num | Avg Lat | Avg TTFT | Avg TPOT | Avg Thpt | Avg In | Avg Out |
|---|-----|------|---------|---------|---------|----------|------|-----|
| DeepSeek-V4-Flash| GPQA-Diamond | 198 | 149.111 s | 1943.3 ms | 14.3 ms | 67 tok/s | 344 | 9991 |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35696480866](https://github.com/sgl-project/sglang/actions/runs/35696480866)
Latest PR Test (Extra): :x: [Run #35696480321](https://github.com/sgl-project/sglang/actions/runs/35696480321)
Latest PR Test (AMD ROCm 10): :hourglass_flowing_sand: [Run #35696480550](https://github.com/sgl-project/sglang/actions/runs/35696480550)

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v4.py [hunk 1]
```diff
@@ -750,11 +750,13 @@ def __init__(
             tp_size=self.attn_tp_size,
             **({} if quantize_wo_a else {"params_dtype": torch.bfloat16}),
         )
-        if quantize_wo_a:
+        if fp8:
```
Comment (AndyLi429) [inline python/sglang/srt/models/deepseek_v4.py]: why change this?

Comment (randgun) [reply python/sglang/srt/models/deepseek_v4.py]: wo_a在dspark层没量化

File: python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py [hunk 2]
```diff
@@ -142,7 +142,16 @@ def sampler(step_logits: torch.Tensor, step_idx: int) -> torch.Tensor:
                     del step_idx
                     # In-graph philox noise: each replay advances the generator
                     # and redraws.
-                    noise = self.exp_noise[:bs].exponential_()
+                    from sglang.srt.utils import is_npu
+
+                    if is_npu():
```
Comment (randgun) [inline python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py]: Only npu arch35 do not support exponential_()

File: python/sglang/srt/models/deepseek_v4_dspark.py [hunk 3]
```diff
@@ -782,15 +783,15 @@ class DeepseekV4ForCausalLMDSpark(nn.Module):
...
+    uses_own_vocab_modules = _is_npu and not is_npu_arch35()
 
     @classmethod
     def shared_experts_fusion_disable_reason(
         cls,
         hf_config,
         quant_config,
     ):
-        if _is_npu:
+        if _is_npu and not is_npu_arch35():
```
Comment (randgun) [inline python/sglang/srt/models/deepseek_v4_dspark.py]: replace to uses_own_vocab_modules

File: python/sglang/srt/models/deepseek_v4.py [hunk 4]
```diff
@@ -883,7 +885,7 @@ def __init__(
             assert hasattr(self.wo_a, "weight_scale_inv"), (
                 "FP8 quant_config must create weight_scale_inv"
             )
-        if self.use_npu_arch35_mxfp8_wo_a:
+        if self.use_npu_arch35_mxfp8_wo_a and hasattr(self.wo_a, "weight_scale_inv"):
```
Comment (randgun) [inline python/sglang/srt/models/deepseek_v4.py]: revert
